Source/GSStoryboardTransform.m

```diff
@@ -0,0 +1,871 @@
+    }
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  // this is never encoded directly...
+}
+
+- (id) instantiate
+{
+  NSStoryboard *sb = [NSStoryboard storyboardWithName: _storyboardName
+                                               bundle: [NSBundle mainBundle]];
+  return [sb instantiateInitialController];
+}
+
+@end
+
+@implementation GSStoryboardTransform
+
+- (instancetype) initWithData: (NSData *)data
+{
+  self = [super init];
+  if (self != nil)
+    {
+      NSXMLDocument *xml = [[NSXMLDocument alloc] initWithData: data
+                                                       options: 0
+                                                         error: NULL];
+      
+      _scenesMap = [[NSMutableDictionary alloc] initWithCapacity: 10];
+      _controllerMap = [[NSMutableDictionary alloc] initWithCapacity: 10];
+      _identifierToSegueMap = [[NSMutableDictionary alloc] initWithCapacity: 10];
+      
+      [self processStoryboard: xml];
+      RELEASE(xml);
+    }
+  return self;
+}
+
+- (void) dealloc
+{
+  RELEASE(_initialViewControllerId);
+  RELEASE(_applicationSceneId);
+  RELEASE(_scenesMap);
+  RELEASE(_controllerMap);
+  RELEASE(_identifierToSegueMap);
+  [super dealloc];
+}
+
+- (NSString *) initialViewControllerId
+{
+  return _initialViewControllerId;
+}
+
+- (NSString *) applicationSceneId
+{
+  return _applicationSceneId;
+}
+
+- (NSMapTable *) segueMapForIdentifier: (NSString *)identifier
+{
+  return [_identifierToSegueMap objectForKey: identifier];
+}
+
+- (NSXMLElement *) createCustomObjectWithId: (NSString *)ident
+                                   userLabel: (NSString *)userLabel
+                                 customClass: (NSString *)className
+{
+  NSXMLElement *customObject =
+    [[NSXMLElement alloc] initWithName: @"customObject"];
+  NSXMLNode *idValue =
+    [NSXMLNode attributeWithName: @"id"
+                     stringValue: ident];
+  NSXMLNode *usrLabel =
+    [NSXMLNode attributeWithName: @"userLabel"
+                     stringValue: userLabel];
+  NSXMLNode *customCls =
+    [NSXMLNode attributeWithName: @"customClass"
+                     stringValue: className];
+  
+  [customObject addAttribute: idValue];
+  [customObject addAttribute: usrLabel];
+  [customObject addAttribute: customCls];
+
+  AUTORELEASE(customObject);
+  
+  return customObject;
+}
+
+- (NSData *) dataForIdentifier: (NSString *)identifier
+{
+  NSString *sceneId = [_controllerMap objectForKey: identifier];
+  NSXMLDocument *xml = [_scenesMap objectForKey: sceneId];
+  return [xml XMLData];
+}
+
+- (NSData *) dataForApplicationScene
+{
+  return [self dataForIdentifier: APPLICATION];
+}
+
+- (void) processChild: (NSXMLElement *)child
+              withDoc: (NSXMLElement *)doc
+          withAppNode: (NSXMLNode *)appNode
+              sceneId: (NSString *)sceneId
+{
+  NSString *customClassString = nil;
+  if (appNode != nil)
+    {
+      NSXMLElement *objects = (NSXMLElement *)[appNode parent];
```

Comment:
The node "objects" will always be the same node as "child". The code here would be a lot easier to understand if you handed hin "child" with the name "objects" and used it consistently

Reply:
I have refactored this to use your code.  There is a difference here.  When the <application> node is present, the <objects> node appears within it, so it's necessary to pull that out.  But it does reside in a different place.

Reply:
Actually I was thinking more in the line of this structure:

`- (void) processChild: (NSXMLElement *)objects
               withDoc: (NSXMLElement *)doc
           withAppNode: (NSXMLNode *)appNode
               sceneId: (NSString *)sceneId
 {
   NSString *customClassString = nil;
   NSXMLNode *appCons = nil;

   if (appNode != nil)
     {
       NSArray *appConsArr = [appNode nodesForXPath: @"connections" error: NULL];

       appCons = [appConsArr objectAtIndex: 0];
       if (appCons != nil)
         {
           [appCons detach];
         }

       // Assign application scene...
       ASSIGN(_applicationSceneId, sceneId);
       [_controllerMap setObject: _applicationSceneId
                          forKey: APPLICATION];

       // Move all application children to objects...
       NSArray *appChildren = [appNode children];      
       FOR_IN(NSXMLElement*, ae, appChildren)
         {
           [ae detach];
           [objects addChild: ae];
         }
       END_FOR_IN(appChildren);

       // Remove the appNode
       [appNode detach];

       // create a customObject entry for NSApplication reference...
       NSXMLNode *appCustomClass = [(NSXMLElement *)appNode
                                                    attributeForName: @"customClass"];
       customClassString = ([appCustomClass stringValue] == nil) ?
         @"NSApplication" : [appCustomClass stringValue];
     }

     [self addStandardObjects: objects
                    classString: customClassString
                    connections: appCons];

     // Add it to the document
     [objects detach];
     [doc addChild: objects;
 }`

Removing a code duplication and an unneeded type cast along the way.

Reply:
This is done.